src/js/ProviderManager.js

```diff
@@ -0,0 +1,45 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*globals async, DirectoryLinksProvider, Links, PlacesProvider, gUserDatabase, gPinnedLinks, gBlockedLinks*/
+
+/**
+ * Singleton that handles link providers.
+ */
+
+ "use strict";
+
+const ProviderManager = {
+  init: async(function*(gMockFetch) {
```

Comment:
@oyiptong @emtwo, test wise, it seems totally possible to either: 
1. rig up Karam to spit out JSON responses on `http://localhost:9999/whatever`
2. Make the SW do the above via the cache API. Should also be fairly trivial to do.  

I've be happy with any of those and happy to help make either happen.